### PR TITLE
Sync3n Initial Add

### PR DESCRIPTION
**Pull request by garrettwrong**
See #1107 .  This adapts some of the CL code brought over for Cn molecules which used the 3N sync matrix method to the C1 case.

Todo
- [x] Add S weighting, as was optional in the Matlab code.
- [x] Add J weighting, as was optional in the Matlab code.
- [x] Speedup, its borderline unusable for experimental runs and gives little feedback...
- [x] Docstrings
- [x] Tests
- [ ] Refactor with other 3N based classes.  Probably they (Cn) could inherit from this Sync3N one.

I quickly hacked this together enough to minimally function. I'm thinking to take on the 1-3 todo items while Josh is out and pass it along to him to finish integrating.

### CODE REVIEW

**Comment by garrettwrong** [comment]
Adding some notes.

I've created a CUDA kernel for the slow operation `signs_times_v` and minimal (cupy) launching code along with some simple auto enabling configuration.  The speedups are 100x+ for relevant problems.  Presently I am keeping that work in this branch (off of sync3n): https://github.com/ComputationalCryoEM/ASPIRE-Python/tree/sync3n_cupy 

There are some concerns/bugs between the python, matlab, and mex versions that need to be addressed, but they would be minor changes.  Josh might be working them out in his current D2 work now anyway.

The original Mex codes of this thead-loop structure all appear to have a common threading ABA bug... I have fixed that in my CUDA version.

I will  be able to mimic the kernel and supporting code I wrote for the other similar mex codes (probability and triangle scores for the same speedups, but need to work on a few other projects first before I come back to that.

**Comment by garrettwrong** [comment]
Some updates.

Today I was able to complete a rough draft of the `pairs_probabilities` CUDA kernel, cupy launcher, and dispatch code for that function.  Speedups are similar to the last kernel 100x+.  Next is the `triangle_scores` kernel which should be similar.

After that I can move onto merging the `_cupy` branch into this branch and cleaning things up.  With the accelerated code we should be able to actually test the Sync3N features.  FWIW, the large experiment using the pure python/numpy sync3n implementation was kicked off before I started writing the kernel code... and the handedness synchronization was only half way done yesterday...

**Comment by garrettwrong** [comment]
For 1000 images ...
```
Size:	1000
Allclose inds?  True
Allclose arb?  True
gpu_time: 7.763748619996477
host_time: 8093.168460645
speedup: 1042.4305135022098
```

**Comment by garrettwrong** [comment]
Triangle scores is also fast now:

```
Size:   1000
Allclose cum_scores?  True
Allclose hist_scores?  True
gpu_time: 4.986900245989091
host_time: 6058.410942732007
speedup: 1214.8650752749106
```

Added a test file to compare between the host python and cupy launched CUDA codes.

Broke the CUDA codes out into their own source file so that it is easier to keep it clean.

Merged sync3n_cupy into this branch.

**Comment by garrettwrong** [comment]
Fixed up some typing concerns.

Added some minimal docstrings.

Added a unit test that conditionally compares the host (numpy) implementation to gpu (cupy launched cuda).  This test is coded to automatically skip when `cupy` is not able to be imported.

**Comment by garrettwrong** [comment]
Added 3 tests to compare dummy data with results from MATLAB as a sanity check of the low level function port.

Added some minimal up casting to allow running the underlying sync3n methods in both singles and doubles.  This is tested with parameterized fixture.  Would still recommend using doubles with our CL codes for now.

**Comment by garrettwrong** [comment]
Added a unit test based on CL Sync 2N and tried to complete most of the docstrings.

Found a small bug in the CL sync tests (doesn't appear it was actually testing both types, just running doubles twice).

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: defaulting to numpy

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: Looks like a note to self. Is the axis good here?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: representing

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: Can clean up to `np.diag(np.kron(np.diag(Dhalf), np.ones(3)))`

Comment (j-c-c) [reply src/aspire/abinitio/commonline_sync3n.py]: Or even cleaner: `np.kron(Dhalf, np.diag(np.ones(3)))`

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: Should we specify dtype here?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py:392]: This indexing should be good for "all triplets". It could be more explicit with something like:
```python
>>> n_img = 5
>>> for i in range(n_img - 2):
...     for j in range(i + 1, n_img-1):
...             for k in range(j + 1, n_img):
...                     print(i,j,k)
... 
0 1 2
0 1 3
0 1 4
0 2 3
0 2 4
0 3 4
1 2 3
1 2 4
1 3 4
2 3 4
```

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: Same suggestion as above for the triple loop.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: Is this comment still relevant?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: May want to change this to `_estimate_relative_rotations()` (and the docstring) or possibly get rid of this method altogether and just call `_estimate_all_Rijs(self.clmatrix)` from the main `estimate_rotations()` method.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [inline src/aspire/abinitio/commonline_sync3n.py]: Same remark as above for all triplet loop

File: tests/test_orient_sync_voting.py
Comment (j-c-c) [inline tests/test_orient_sync_voting.py:55]: Oops. Thanks!

**Review by j-c-c — CHANGES_REQUESTED** [review]
This is great! Just a few things.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: updated

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Yeah, was a note to myself. Should be. I'll remove.

```
% Prepare weights for the blocks of the matrix S.
D = mean(W,2); % We use mean and not sum so that after normalizing W the 
               % sum of its rows will be N, as required.
```

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: updated

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: I'll take the first one, oversight on my part translating the code, but the second one is not the same...

```
>>> Dhalf = np.arange(9).reshape(3,3)
>>> a=np.diag(np.kron(np.diag(Dhalf), np.ones((1, 3)))[0])
>>> b=np.diag(np.kron(np.diag(Dhalf), np.ones(3)))
>>> c=np.kron(Dhalf, np.diag(np.ones(3)))
>>> 
>>> a
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> b
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> c
array([[0., 0., 0., 1., 0., 0., 2., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 2., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 2.],
       [3., 0., 0., 4., 0., 0., 5., 0., 0.],
       [0., 3., 0., 0., 4., 0., 0., 5., 0.],
       [0., 0., 3., 0., 0., 4., 0., 0., 5.],
       [6., 0., 0., 7., 0., 0., 8., 0., 0.],
       [0., 6., 0., 0., 7., 0., 0., 8., 0.],
       
 ```

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: I added `self.dtype` for W and ones arrays that construct D.  I fear most of this code probably doesn't work well in singles.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:392]: I'm not sure what you mean here, because your suggection makes the same sequence.

Is it that you just dont like the two empty loops reulting from the outer `range(n_img)` vs `range(n_img-2)`? I'm assuming that's it and updated, but if it was something else lmk.

```
>>> n_img = 5
>>> for i in range(n_img - 2):
...   for j in range(i + 1, n_img-1):
...      for k in range(j + 1, n_img):
...        print(i,j,k)
... 
0 1 2
0 1 3
0 1 4
0 2 3
0 2 4
0 3 4
1 2 3
1 2 4
1 3 4
2 3 4
>>> for i in range(n_img):
...   for j in range(i + 1, n_img-1):
...      for k in range(j + 1, n_img):
...        print(i,j,k)
... 
0 1 2
0 1 3
0 1 4
0 2 3
0 2 4
0 3 4
1 2 3
1 2 4
1 3 4
2 3 4
```

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: added the `-2` to avoid the no-op iterations.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: maybe, i just removed the `peak` var; wasn't adding much to the code.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: `-2`'d it

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Sure, I just moved the functional lines into `estimate_rotations`

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [reply src/aspire/abinitio/commonline_sync3n.py]: Sounds good. Third one does work for an already diagonal `Dhalf`.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [reply src/aspire/abinitio/commonline_sync3n.py:392]: There was a comment about checking the bound. I was just saying that the bound is good and giving a suggestion if you wanted to make the bounds explicit (ie. no empty loops).

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Oh, I see so `d=p.kron(np.diag(Dhalf), np.diag(np.ones(3))). I'll try it

Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Nah, still didn't work. Going to stick with the one referred to as `b` above.  That removes the silly oversight I had in there. Thanks for catching.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:392]: Okay cool.  Yeah, I didn't like those bounds when I ported them, but maybe they were the secret sauce 🤷‍♂️  lol.  Thanks for double checking.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [reply src/aspire/abinitio/commonline_sync3n.py]: Sounds good!

But just to clarify, `Dhalf` in the code is already a diagonal matrix. Your example starts with a non-diagonal `Dhalf`. If you start with `Dhalf` diagonal `a`, `b`, and `c` should be the same.

Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: See above. I've tried both ways (c, then d), neither worked 🤷‍♂️ .

Comment (j-c-c) [reply src/aspire/abinitio/commonline_sync3n.py]:
```
>>> Dhalf = np.diag(np.diag(np.arange(9).reshape(3,3)))
>>> Dhalf
array([[0, 0, 0],
       [0, 4, 0],
       [0, 0, 8]])
>>> a=np.diag(np.kron(np.diag(Dhalf), np.ones((1, 3)))[0])
>>> b=np.diag(np.kron(np.diag(Dhalf), np.ones(3)))
>>> c=np.kron(Dhalf, np.diag(np.ones(3)))
>>> a
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> b
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> c
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
```

Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: I would be super caution about injecting implicit behavior here.
These are _clearly_ not doing the same thing outside a special case...  While it may be the one used in that line today, if any other line is changed, yikes.... let's not.

>>> Dhalf = np.arange(9).reshape(3,3)
>>> d=np.kron(np.diag(Dhalf), np.diag(np.ones(3)))                                             
>>> d                                                                                          
array([[0., 0., 0., 4., 0., 0., 8., 0., 0.],                                                   
       [0., 0., 0., 0., 4., 0., 0., 8., 0.],                                                   
       [0., 0., 0., 0., 0., 4., 0., 0., 8.]])

**Comment by garrettwrong** [comment]
Going to wait until after `hack24` is reviewed/merged to retest and move this forward.  (They both touch/have gpu/cupy code.)

File: src/aspire/abinitio/commonline_sync3n.py
Comment (j-c-c) [reply src/aspire/abinitio/commonline_sync3n.py]: fair enough!

**Comment by garrettwrong** [comment]
Running with JSB2017 Matlab class averages.

**Comment by garrettwrong** [comment]
> Running with JSB2017 Matlab class averages.

For the 80s run at 89 pixels the Sweighting solve was infeasable (I'll look into that more, happens too much).  Without Sweighting we did produce a reasonable molecule. Average aligned rotation error was ~16* wrt MATLAB, FSC was 80A wrt to EMDB.  Not very good, but better than before...

I noticed MATLAB was run using twice as many `n_theta` and must also have been using non-default shift steps.  I am retrying with those changes.  Unfortunately I have not found the actual configuration used in the MATLAB logs yet... still looking.  I'm basing the values on lines from areas of the log, but its not certain they were the same values for the different functions.  Anyway, hoping this will at least yield some improvement.

**Comment by garrettwrong** [comment]
Rebased with lated `develop`  and added a commit (oversight) I found useful doing upstream work.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py]: resolution → box size? I know we use “resolution” in many places to mean this, but it becomes especially confusing in places like this where “resolution” is used just below to mean something else. We should probably aim to use this language throughout but let's start with new additions.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py:214]: Wouldn't `np.kron(Dhalf, np.ones((3, 3), dtype=self.dtype))` work here?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py]: It's `W`, though, right?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py:389]: I forget what happens to `trange`s in our code. Can the user suppress them if needed or do they always output. I know with logging, we have that option, but what about for `tqdm` and friends?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py]: This is a bit confusing. Can we use `np.argmax(s_ik_jk < h * np.arange(…))` or something here?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py:566]: Looks like up to this point, we have the same code as `_triangle_scores_inner_host`. Would it make sense to factor it out into its own function?

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [inline src/aspire/abinitio/commonline_sync3n.py:115]: Should we have a way to force CPU even if GPU is present (maybe the user has a faulty device or something, although one would expect `import cupy` to fail there)?

**Review by janden — CHANGES_REQUESTED** [review]
Thank you for putting this together and sorry that it took so long for me to review it. This all looks very good and there are just a few things to discuss, then we can get this in!

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:115]: Uninstalling `cupy` would do it, but that is a little roundabout. Added a `disable_gpu` flag.  

Keep in mind this algorithm is practically unusable beyond tiny problem on the CPU.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:389]: We have wrapped `trange` and `tqdm` to be controlled (suppressible) by our configuration.  I think there is an example in our docs gallery.  We just need to remember to import from our utils, which I've done here:

`from aspire.utils import J_conjugate, all_pairs, nearest_rotations, tqdm, trange`

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Yeah, my bad.  Changed.

They're basically the same; P is the flattened triangle which is expanded into a symmetric matrix for W.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Yeah I just copied these from our existing CL code. We almost have the complete set of CL codes at which point it might make sense to refactor them.

Changed:

```
        :param max_shift: Maximum range for shifts as a proportion of box size. Default = 0.15.                 
        :param shift_step: Step size of shift estimation in pixels. Default = 1 pixel.
```

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Probably I could just use division and a clamp... I'll change it

Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py]: Rather a multiplication 

```
                    # Compute scores                                                                            
                    s[0] = 1 - np.sqrt(best_val / alt_ij_jk)  # s_ij_jk                                         
                    s[1] = 1 - np.sqrt(best_val / alt_ik_jk)  # s_ik_jk                                         
                    s[2] = 1 - np.sqrt(best_val / alt_ij_ik)  # s_ij_ik                                         

                    # Update histogram                                                                          
                    # Find integer bin [0,self.hist_intervals)                                                  
                    _l1, _l2, _l3 = np.minimum(
                        (self.hist_intervals * s).astype(int),  # implicit floor                                
                        self.hist_intervals-1)  # clamp upper bound                                             

                    scores_hist[_l1] += 1
                    scores_hist[_l2] += 1
                    scores_hist[_l3] += 1
```

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:214]: Nope.  
```
            # Make Dhalf of size 3Nx3N                                                                          
            _Dhalf = np.kron(Dhalf, np.ones(3, 3), dtype=self.dtype))
            Dhalf = np.diag(np.kron(np.diag(Dhalf), np.ones(3, dtype=self.dtype)))
            np.testing.assert_allclose(_Dhalf, Dhalf)  # fails  
```          

Josh did suggest an alternative that would work in this area, but it was only equivalent based on other lines of code far above, and I have concerns that is setting a trap should those lines ever change....

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:566]: They are very similar but how/which various vars are used is slightly different.  I did successfully factor it out, but it isn't clear to me whether it is cleaner or just more abstract.  

These loops are super time consuming on cpu... to avoid constantly initializing arrays I pass them by ref (which is potentially confusing) instead of functionally returning them.  It also is getting farther from following the MATLAB code, for whatever that is worth.

Its a single commit so I can revert it if needed. LMK.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py]: Thanks!

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py:214]: Huh. Never mind then…

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py]: Ok got it. Thanks!

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py:389]: Ok I see.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py]: Much cleaner. Thanks!

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py:115]: > Uninstalling `cupy` would do it, but that is a little roundabout. Added a `disable_gpu` flag.

Ok sounds good.

> Keep in mind this algorithm is practically unusable beyond tiny problem on the CPU.

Yeah I know, but it's just good to have the option.

File: src/aspire/abinitio/commonline_sync3n.py
Comment (janden) [reply src/aspire/abinitio/commonline_sync3n.py:566]: No that's fine. Let's just keep as is. I didn't realize there were some differences.

**Review by janden — DISMISSED** [review]

File: src/aspire/abinitio/commonline_sync3n.py
Comment (garrettwrong) [reply src/aspire/abinitio/commonline_sync3n.py:566]: Ok I'll revert that commit then merge.

**Pull request merged**
